heir_py/mlir_emitter.py

```diff
@@ -25,38 +116,60 @@ def emit(self):
     # TODO(#1162): get inferred or explicit return types
     return_types_str = "i64"
 
-    body = self.emit_body()
+    body = self.emit_blocks()
 
     mlir_func = f"""func.func @{func_name}({args_str}) -> ({return_types_str}) {{
 {textwrap.indent(body, '  ')}
 }}
 """
     return mlir_func
 
-  def emit_body(self):
+  def emit_blocks(self):
     blocks = self.ssa_ir.blocks
-    str_blocks = []
-    first = True
 
+    # collect loops and block header needs
+    self.omit_block_header[sorted(blocks.items())[0][0]] = True
     for block_id, block in sorted(blocks.items()):
-      instructions = []
-      for instr in block.body:
-        result = self.emit_instruction(instr)
-        if result:
-          instructions.append(result)
+      for i in range(len(block.body)):
+        # Detect a range call
+        instr = block.body[i]
+        if is_range_call(instr, self.ssa_ir):
```

Comment:
This assumes that the only place `range` is used is within a for loop. A decent simplifying assumption for now, but we might want to tighten that a little bit to something like "is_start_of_static_loop(block)" that checks for the range and the subsequent instructions in the block to match it as a loop header. Otherwise the failure mode here is an assertion failure while attempting to build a loop.